### ISSUE TITLE
QuickBasic calculation error

### ISSUE THREAD

**Issue by akcansoft** [joncampbell123/dosbox-x#1939]
Quickbasic calculation error on DOSBOX-X

Dosbox-X v0.83.6
`PRINT VAL("5")`
**OUTPUT:**
QuickBasic v4.5 -> **4.999999999999999**
Qbasic v1.1 -> **5**
QBX v7.1 -> **5**

DOSBOX v0.74-3
`PRINT VAL("5")`
**OUTPUT:**
QuickBasic v4.5 -> **5**

**Comment by Wengier** [comment]
I have just tried QuickBasic 4.5 and run the command ```PRINT VAL("5")``` inside DOSBox-X 0.83.6, and this is the result I got:

![image](https://user-images.githubusercontent.com/8216923/95666208-6ddfa180-0b25-11eb-8288-07d66bd0aeba.png)

![image](https://user-images.githubusercontent.com/8216923/95666250-daf33700-0b25-11eb-84e9-ff4c2be77497.png)

I downloaded QuickBasic 4.5 from:

https://winworldpc.com/product/quickbasic/45

So it seems that the result here is correct (5 instead of 4.999999999999999). I tried with the default configuration (```dosbox-x.exe -defaultconf```). If you have modified the configuration file, you can certainly try with the default setting and see the result.

**Comment by joncampbell123** [comment]
If you are using VS2019 builds and NOT using dynamic core, you may get some loss of floating point precision. VS2019 does not support the "long double" data type.

The same issue exists with a QBASIC game that checks it's own EXE file size.

Try running a MinGW build, or switch on dynamic core, or use dosbox.conf to disable the FPU so that QBASIC is forced to emulate floating point in software.

**Comment by akcansoft** [comment]
I tried with the default configuration.
`dosbox-x.exe -defaultconf`
not modified DOSBox-X configuration file.
QuickBasic v4.5
`PRINT VAL("5")`
Result : **4.999999999999999**
![ss](https://user-images.githubusercontent.com/7031346/95713676-a25f7480-0c6f-11eb-8d63-48b9e6b067de.png)

**Comment by Wengier** [comment]
@akcansoft Can you please post a screenshot that shows the full welcome banner ("Welcome to DOSBox-X ! ..."), which will include details about the actual build you are using, e.g. Windows (VS) or MinGW, 32-bit or 64-bit, SDL1 or SDL2, etc? Thanks!

**Comment by akcansoft** [comment]
i tried new download and install
download page: https://github.com/joncampbell123/dosbox-x/releases
download file:  dosbox-x-vsbuild-win-20200930224615.zip 
https://github.com/joncampbell123/dosbox-x/releases/download/dosbox-x-v0.83.6/dosbox-x-vsbuild-win-20200930224615.zip
extact zip file to folder
run: folder/bin/x64/Release/dosbox-x.exe
SS:
![command_000](https://user-images.githubusercontent.com/7031346/95717873-4b10d280-0c76-11eb-8347-33f3293421e9.png)
mount c  c:\dosprg
c:
cd QB45
QB
Quickbasic v4.5
`PRINT VAL("5")` Result: 4.999999999999999

run: folder\x64\Release SDL2\dosbox-x.exe
result is the same

next. new test:
download and unzip file: dosbox-x-mingw-win64-20200930235023.zip
https://github.com/joncampbell123/dosbox-x/releases/download/dosbox-x-v0.83.6/dosbox-x-mingw-win64-20200930235023.zip
folder\mingw-build\mingw\dosbox-x.exe
qb test result : 5
it's ok.
should I use this version?

**Comment by Wengier** [comment]
@akcansoft Yes, the last version is the MinGW build that @joncampbell123 mentioned earlier, which should work fine for this case.

**Comment by blindphoneman** [comment]
Interestingly, I ran QB under DOS 6.22 running in an HD image under DOSBox-x and it gave 5 as the answer.

**Comment by Dwedit** [comment]
Is this about the initial state of Floating Point rounding mode?

**Comment by Dwedit** [comment]
I think this issue is serious enough that the either MSVC builds should have the Dynamic core selected by default, or the MinGW builds should be prioritized by the installer instead of the MSVC builds.

**Comment by Dwedit** [comment]
Bug is still present in newest release.  Maybe make the MingW build the default?

**Comment by joncampbell123** [comment]
Can you try adding:

[cpu]
fpu=false

**Comment by Dwedit** [comment]
Tried logging FPU related stuff, only saw this in the logfile:
```
         0       FPU:FPU core: double FPU (caution: possible precision errors)
```

And I am aware that picking the MinGW build fixes the problem, but the MinGW build is not the preselected build when you try to install the program.

**Comment by Allofich** [comment]
I haven't really looked into it, but when I tried the MinGW build recently, it seemed to have a problem with delayed input in the game Descent. I tried the Visual Studio build after that (built on the same code) and it didn't have the problem. So the MinGW build may have its own issues and isn't necessarily the superior option of the two.
